processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -305,125 +300,75 @@ public int seek(int dictId)
+          mergers.get(i).convertSegmentRowValuesToMergedRowValues(indexNumber, sourceDimensionSelector);
+      convertedDimensionSelectors[i] = convertedDimensionSelector;
+      //noinspection ObjectEquality
+      anySelectorChanged |= convertedDimensionSelector != sourceDimensionSelector;
+
+      convertedMarkedDimensionSelectors[i] = mergers.get(i).convertSegmentRowValuesToMergedRowValues(
+          indexNumber,
+          sourceRowIterator.getMarkedPointer().dimensionSelectors[i]
       );
     }
-  }
-
-  class RowboatMergeFunction implements BinaryFn<Rowboat, Rowboat, Rowboat>
-  {
-    private final AggregatorFactory[] metricAggs;
-
-    public RowboatMergeFunction(AggregatorFactory[] metricAggs)
-    {
-      this.metricAggs = metricAggs;
+    if (!anySelectorChanged) {
+      return sourceRowIterator;
     }
-
-    @Override
-    public Rowboat apply(Rowboat lhs, Rowboat rhs)
+    RowPointer convertedRowPointer = sourceRowPointer.withDimensionSelectors(
+        convertedDimensionSelectors,
+        sourceRowPointer.getDimensionHandlers()
+    );
+    TimeAndDimsPointer convertedMarkedRowPointer = sourceRowIterator.getMarkedPointer().withDimensionSelectors(
+        convertedMarkedDimensionSelectors,
+        sourceRowIterator.getMarkedPointer().getDimensionHandlers()
+    );
+    return new ForwardingRowIterator(sourceRowIterator)
     {
-      if (lhs == null) {
-        return rhs;
-      }
-      if (rhs == null) {
-        return lhs;
+      @Override
+      public RowPointer getPointer()
+      {
+        return convertedRowPointer;
       }
 
-      Object[] metrics = new Object[metricAggs.length];
-      Object[] lhsMetrics = lhs.getMetrics();
-      Object[] rhsMetrics = rhs.getMetrics();
-
-      for (int i = 0; i < metrics.length; ++i) {
-        Object lhsMetric = lhsMetrics[i];
-        Object rhsMetric = rhsMetrics[i];
-        if (lhsMetric == null) {
-          metrics[i] = rhsMetric;
-        } else if (rhsMetric == null) {
-          metrics[i] = lhsMetric;
-        } else {
-          metrics[i] = metricAggs[i].combine(lhsMetric, rhsMetric);
-        }
-      }
-
-      final Rowboat retVal = new Rowboat(
-          lhs.getTimestamp(),
-          lhs.getDims(),
-          metrics,
-          lhs.getRowNum(),
-          lhs.getHandlers()
-      );
-
-      for (Rowboat rowboat : Arrays.asList(lhs, rhs)) {
-        Iterator<Int2ObjectMap.Entry<IntSortedSet>> entryIterator =
-            rowboat.getComprisedRows().int2ObjectEntrySet().fastIterator();
-        while (entryIterator.hasNext()) {
-          Int2ObjectMap.Entry<IntSortedSet> entry = entryIterator.next();
-
-          for (IntIterator setIterator = entry.getValue().iterator(); setIterator.hasNext(); /* NOP */) {
-            int rowNum = setIterator.nextInt();
-            retVal.addRow(entry.getIntKey(), rowNum);
-          }
-        }
+      @Override
+      public TimeAndDimsPointer getMarkedPointer()
```

Comment:
please add javadocs for getPointer and getMarkedPointer

Reply:
Those methods are documented on interface level: see `TimeAndDimsIterator`, `RowIterator`, `TransformableRowIterator`. Let me know if something is unclear from those specifications.